### PR TITLE
[APP] ICR FE Integration

### PR DESCRIPTION
**Pull request by psparacino**
## Description

Carbonmark Integration for ICR credits

**Notes for reviewers:**

1) The retirement data for receipts in dependent on merge of #2048 . This will eventually all be moved to an endpoint.
2) `serialization` is one of three possible query parameters for an individual on the ICR api. `serialization` is used over the other two options (`contractAddress` and ICR's version of a `projectID`) whenever possible for consistency.
3) ICR quantities are regular integers, so the amounts are handled separately for the ICR registry [here](https://github.com/KlimaDAO/klimadao/compare/APP_feature/icr-api-integration?expand=1#diff-1a13cfc031e74278375852de1ca0bf4b8d421b8ed5125f837cd8cc799016d9dbR62). 
4) This PR uses a mainnet testing setup: mainnet RA, carbonmark contract, ICR Carbon Registry contract, ICR subgraph, digital-carbon and marketplace subgraphs. These all need to be changed to production versions before merge.
5) The API is currently hardcoded to the API PR Preview #2051 . This should be [removed](https://github.com/KlimaDAO/klimadao/pull/2052/files#diff-a1be0ef50959b412feee17a3bcc720124c83ccf3e1ab62f240d5ad0d669a1aa4R38) before merge.

To change before merge:
- [x] AggregatorV2 contract
- [x] Carbonmark contract 
- [x] ICR Carbon Registry contract
- [x] ICR subgraph
- [x] Digital-carbon subgraph
- [x] Marketplace subgraph
- [x] Hardcoded Preview URL

Other improvements were noted along the way that weren't necessarily ICR specific but were rolled into this PR:
1) Removing duplicate prices or quantities when only editing one option in a listing: [here](https://github.com/KlimaDAO/klimadao/compare/APP_feature/icr-api-integration?expand=1#diff-7c25eb7f0d621afca61781b09f378ef2b304e011f6bdfe6a1b8785d2b1c616b5R92)
2) Fix recurring nesting error because of title construction in Text component [here](https://github.com/KlimaDAO/klimadao/compare/APP_feature/icr-api-integration?expand=1#diff-4d85183ccd99ed8e794114df235358835b17f2e288ac462bda3c5625383179c0R37)

## Related Ticket

Also related to #2051 

## Notes For QA

Specific pages, components or journeys that might be affected:
- All need to be tested

Relevant preview URLs:
- 

Other notes:
- This is a large PR(!). Please feel free to reach out for any clarification or to pair review any specific section.

### CODE REVIEW

File: carbonmark/lib/constants.ts [hunk 1]
```diff
@@ -22,21 +22,26 @@ export const WALLETCONNECT_PROJECT_ID =
...
-);
+// @todo uncomment before merge
+// const SHORT_COMMIT_HASH = process.env.NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA?.slice(
+//   0,
+//   7
+// );
 
 /** When incrementing this API version, be sure to update TypeScript types to reflect API changes */
-export const API_PROD_URL = "https://v5.0.0.api.carbonmark.com";
+export const API_PROD_URL = "https://v5.1.1.api.carbonmark.com";
```
Comment (biwano) [inline carbonmark/lib/constants.ts:32]: This does not match with the API version in the API PR

File: carbonmark/lib/createUrls.ts [hunk 2]
```diff
@@ -1,11 +1,20 @@
...
   key: Project["key"];
   vintage: Project["vintage"];
+  serialization?: Project["serialization"] | null;
+};
+
+export const createProjectLink = (projectData: ProjectData) => {
+  if (projectData.key.startsWith("ICR") && projectData.serialization) {
+    return `/projects/${projectData.serialization}`;
+  } else {
+    return `/projects/${projectData.key}-${projectData.vintage}`;
```
Comment (biwano) [inline carbonmark/lib/createUrls.ts:15]: this works if `projectData.key.startsWith("ICR") == true` and there is no `projectData.serialization` ?

File: carbonmark/lib/createUrls.ts [hunk 3]
```diff
@@ -23,3 +32,7 @@ export const createProjectPoolPurchaseLink = (
   project: ProjectData,
   pool: TokenPrice["poolName"]
 ) => `${createProjectLink(project)}/purchase/pools/${pool}`;
+
+export const getPolygonScanBaseUrl = (networkLabel: NetworkParam): string => {
+  return networkLabel === "polygon" ? urls.polygonscan : urls.mumbaiPolygonscan;
+};
```
Comment (biwano) [inline carbonmark/lib/createUrls.ts:38]: Thank you ser again

**Review by biwano — CHANGES_REQUESTED** [review]

File: carbonmark/components/AssetDetails/index.tsx [hunk 4]
```diff
@@ -55,7 +58,12 @@ export const AssetDetails: FC<TotalValuesProps> = (props) => {
         </Text>
       </div>
       <div className={styles.totalsText}>
-        <Anchor className={styles.iconAndText} href={props.polyscanUrl}>
+        <Anchor
+          className={styles.iconAndText}
+          href={`${getPolygonScanBaseUrl(networkLabel)}/address/${
```
[the remark this answers is not in the document]
Comment (0xemc) [reply carbonmark/components/AssetDetails/index.tsx:63]: Let's nuke ` polyscanUrl: string;` from props now too

File: carbonmark/components/CreateListing/Form/index.tsx [hunk 5]
```diff
@@ -92,6 +97,12 @@ export const CreateListingForm: FC<Props> = (props) => {
                   value: Number(selectedAsset.amount),
                   message: t`Balance exceeded`,
                 },
+                validate: {
+                  isWholeNumber: (value) =>
+                    !selectedAsset.id.startsWith("ICR") ||
+                    Number.isInteger(parseFloat(value)) ||
+                    t`ICR credits can only be listed in whole tonnes`,
```
Comment (0xemc) [inline carbonmark/components/CreateListing/Form/index.tsx:104]: Nice pre-emptive catch

File: carbonmark/components/CreateListing/Transaction/index.tsx [hunk 6]
```diff
@@ -58,7 +59,13 @@ export const Transaction: FC<Props> = (props) => {
...
       {view === "approve" && (
         <Approve
-          amount={t`${props.allowance} tonnes`}
+          amount={t`${
+            props.allowance === constants.MaxUint256.toString()
+              ? "Maximum allowance of"
+              : Number(props.allowance) > 10000000
+              ? Number(props.allowance).toExponential(2)
+              : props.allowance
+          } tonnes`}
```
[the remark this answers is not in the document]
Comment (0xemc) [reply carbonmark/components/CreateListing/Transaction/index.tsx:68]: For clarity, this should probably be a labelled formatting function in utils

File: carbonmark/components/CreateListing/index.tsx [hunk 7]
```diff
@@ -96,6 +100,10 @@ export const CreateListing: FC<Props> = (props) => {
    */
   const hasApproval = () => {
     if (!Number(inputValues?.amount)) return false;
+    // 1155 approvals are all or nothing approvals across all tokenIds and amount
+    if (!!inputValues?.tokenId) {
+      return currentAllowance === constants.MaxUint256.toString();
+    }
```
Comment (0xemc) [inline carbonmark/components/CreateListing/index.tsx:106]:
```suggestion
const isEIP1155 = !!inputValues?.tokenId;
    // 1155 approvals are all or nothing approvals across all tokenIds and amount
    if (isEIP1155) {
      return currentAllowance === constants.MaxUint256.toString();
    }
```

File: carbonmark/components/pages/Portfolio/Retire/RetireForm/index.tsx [hunk 8]
```diff
@@ -94,25 +99,38 @@ export const RetireForm = (props: RetireFormProps) => {
...
+    } else if (parts[0].toUpperCase() === "ICR") {
+      return "icr";
     }
     return parts[0].toLowerCase();
   };
 
   const carbonTokenInfo =
     carbonmarkTokenInfoMap[getTokenPrefix(tokenSymbol) as CarbonmarkToken];
 
+  // temporary until digital-carbon subgraph credit its for ICR projects following registry-vintage format
```
Comment (0xemc) [inline carbonmark/components/pages/Portfolio/Retire/RetireForm/index.tsx:111]: Can you create a ticket for this and link it here please?

File: carbonmark/lib/types/carbonmark.types.ts [hunk 9]
```diff
@@ -28,6 +28,8 @@ export type CategoryName = keyof typeof CATEGORY_INFO;
 export interface DigitalCarbonCredit {
   id: string;
   vintage: number;
+  tokenAddress: string;
+  tokenId: string | null;
```
Comment (0xemc) [inline carbonmark/lib/types/carbonmark.types.ts:32]: The sooner we integrate with the new retire endpoints and the sooenr we get rid of these types the better.

Do you have a ticket for integrating with the retire endpoints?

File: carbonmark/components/pages/Users/SellerConnected/Forms/EditListing.tsx [hunk 10]
```diff
@@ -119,6 +120,12 @@ export const EditListing: FC<Props> = (props) => {
                   value: Number(totalAvailableQuantity),
                   message: t`Available balance exceeded`,
                 },
+                validate: {
+                  isWholeNumber: (value) =>
+                    !props.listing.project.id.startsWith("ICR") ||
+                    Number.isInteger(parseFloat(value)) ||
+                    t`ICR credits can only be listed in whole tonnes`,
```
Comment (0xemc) [inline carbonmark/components/pages/Users/SellerConnected/Forms/EditListing.tsx:127]: I've seen this in a few places now, can you please extract it as it's own validator.

File: carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx [hunk 11]
```diff
@@ -61,10 +62,13 @@ export const ListingEditable: FC<Props> = (props) => {
     setIsLoading(true);
     try {
       if (!address) return;
+
       const allowance = await getCarbonmarkAllowance({
         tokenAddress: values.tokenAddress,
         userAddress: address,
         network: networkLabel,
+        // @todo add tokenID to project type on the subgraph. temporary workaround only to mock a tokenID to identify as 1155
```
Comment (0xemc) [inline carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx:70]: Is this still relevant? if so is there a ticket we can link to?

File: carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx [hunk 12]
```diff
@@ -107,6 +115,8 @@ export const ListingEditable: FC<Props> = (props) => {
         signer: provider.getSigner(),
         value: newAllowanceValue,
         onStatus: onUpdateStatus,
+        // @todo add tokenID to project type on the subgraph. temporary workaround only to mock a tokenID to identify as 1155
```
Comment (0xemc) [inline carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx:118]: As above

File: carbonmark/hooks/useFetchUsersWalletOrHandle.ts
Comment (0xemc) [inline carbonmark/hooks/useFetchUsersWalletOrHandle.ts:1]: This has been resolved with the upgrade to kubb2.0

We can delete this file now.

File: carbonmark/lib/createUrls.ts [hunk 13]
```diff
@@ -1,11 +1,20 @@
...
 
 type ProjectData = {
   key: Project["key"];
   vintage: Project["vintage"];
+  serialization?: Project["serialization"] | null;
+};
+
+export const createProjectLink = (projectData: ProjectData) => {
+  if (projectData.key.startsWith("ICR") && projectData.serialization) {
+    return `/projects/${projectData.serialization}`;
```
Comment (0xemc) [inline carbonmark/lib/createUrls.ts:13]: Are we not able to construct an ICR-1234 id?

File: lib/constants/index.ts [hunk 14]
```diff
@@ -72,10 +72,10 @@ const testnet: typeof mainnet = {
   staking: "0x2960DCE5aE04eF503b36f8581EA5Ac5238632092",
   staking_helper: "0x4D70a031Fc76DA6a9bC0C922101A05FA95c3A227",
   retirementStorage: "",
-  retirementAggregatorV2: "",
+  retirementAggregatorV2: "0x62d3897089C93A0fa2B0746A6975Ec4693c13cb8",
   liveOffsetWallet: "",
   liveOffsetContract: "",
-  carbonmark: "0x5dc7083cd6ff9a46c36661750fabcbb9ae2abed2", // v2 Diamond
+  carbonmark: "0xD973F90a4C49607EABeFdb2C45d4F39436c7e7fA", // v2 Diamond
```
Comment (0xemc) [inline lib/constants/index.ts:78]: Dos this v2 Diamond comment need to go?

**Comment by 0xemc** [comment]
Great effort on this one, it's huge. Some nice clean up along the way as well.

A handful of questions & suggestions then good to go after QA and Bruno approve

File: carbonmark/lib/types/carbonmark.types.ts (hunk 9 above)
Comment (psparacino) [reply carbonmark/lib/types/carbonmark.types.ts:32]: https://github.com/KlimaDAO/klimadao/issues/2128

File: carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx (hunk 11 above)
Comment (psparacino) [reply carbonmark/components/pages/Users/SellerConnected/ListingEditable.tsx:70]: I'm not sure there's a clear need for it anymore but it'd definitely be a nice to have for the future. Will create a ticket.

File: carbonmark/lib/constants.ts (hunk 1 above)
Comment (psparacino) [reply carbonmark/lib/constants.ts:32]: I had initially understood the MO to be merge first and then open a separate PR that only increments the API version here? Open to whichever option, just want to confirm

File: carbonmark/components/pages/Portfolio/Retire/RetireForm/index.tsx (hunk 8 above)
Comment (psparacino) [reply carbonmark/components/pages/Portfolio/Retire/RetireForm/index.tsx]: Combined ticket to keep track of these items; will break out post-merge: https://github.com/KlimaDAO/klimadao/issues/2093

**Review by 0xemc — APPROVED** [review]
Good to go, 

Obivously wait for the API PR to close and finish it's release first.

Also uncomment the PREVIEW_URL in `constants.ts`

File: carbonmark/lib/constants.ts (hunk 1 above)
Comment (0xemc) [reply carbonmark/lib/constants.ts]: If we merge first then update staging will be broken in that intermediate time right?

Comment (psparacino) [reply carbonmark/lib/constants.ts]: Just tried with the ICR FE pointing at the current api version and everything seemed(*) to work ok except everything ICR obviously.  Everything else being equal it makes sense to me(?) to just include the version incrementation in the APP PR itself.

And just to confirm the versioning:

Currently FE is on 5.0.0.

Currently API is 5.2.1

ICR API is 5.3.1

So the API PRs and APP PRs would increment both FE and API to 5.3.1. (seems like a big jump for the FE so just want to confirm).

**Review by jabby09 — APPROVED** [review]

**Review by biwano — APPROVED** [review]

**Pull request merged**